File: ANN_lua/annsp.lua
require 'torch'
require 'pl'
--object declaration
local annsp = {}
	
	--inits
	function annsp.init(data, labels)
	    if annsp.verify_dimensions(data) then
	    	annsp.x = data
	    	annsp.y = labels
	    	annsp.eta = 0.01
	    	--annsp.weights = torch.randn(#annsp.x[1]+1)
	    	annsp.weights = torch.Tensor({-0.1839, 0.4486, -0.1335, -0.7360})
	    	annsp.bias = 1
	    	annsp.train()
	    else
	    	print("data not consistent..terminating")
	    end
	end

	--verifies the dimensions and structure of x - if inconsistent, terminate.
	function annsp.verify_dimensions(x)
	    local sum_length = 0
	    local random_pick = #x[torch.random(#x)]
	    for i = 1, #x do
	    	sum_length = sum_length + #x[i]
	    end
	    if sum_length/#x ~= random_pick then
	    	return false
	    else
	    	return true
	    end
	end

	--basic signum (just for demo - can be replaced with either a sigmoid or hyperbolic tangent function)
	function annsp.signum(value)
		if value >= 0 then
			return 1
		else
			return -1
		end
	end

	--adder function
	function annsp.adder(w, x)
		return torch.dot(w,x)
	end

	--adaptation of weight vector - weights updating method
	function annsp.vector_adapt(current_weights, x, expected_decision, decision)
		return current_weights + x*(annsp.eta*(expected_decision-decision))
	end

	--main training function (updates weights on every wrong decision)
	function annsp.train()
		local n_instances = #annsp.x
		--prepending associated class to the vector
		for i=1, n_instances do
			table.insert(annsp.x[i], 1, annsp.y[i])
		end
		annsp.x = torch.Tensor(annsp.x)
		--iterate this many times.
		for l=1, 20 do
			for i=1, n_instances do
				local expected_decision = annsp.y[i]
				local decision = annsp.signum(annsp.adder(annsp.weights, annsp.x[i]))
				if decision ~= expected_decision then
					annsp.weights = annsp.vector_adapt(annsp.weights, annsp.x[i], expected_decision, decision)
					--can add an else condition, which basically will help in terminating the loop over right
					--set of weigts.
				end
			end
		end
	end

	--classifier or predictor function
	function annsp.classify(v)
		local j = v
		table.insert(j, 1, 1)
		j = torch.Tensor(j)
		local decision = annsp.signum(annsp.adder(annsp.weights, j))
		return decision
	end
 
return annsp

File: ANN_py/annsp.py
from __future__ import division
import random
import math
import sys
sys.dont_write_bytecode = True

class annsp(object):
    def __init__(self, data, labels):
        if self.verify_dimensions(data):
            self.x = data
            self.y = labels
            self.weights = [random.uniform(1,len(data)) for _ in range(0, len(data[0])+1)]
            self.eta = 0.01
            self.train()

    def verify_dimensions(self, x):
        random_pick = len(x[random.randrange(1, len(x))])
        if sum(map(len, x))/len(x) == random_pick:
            return True
        else:
            return False

    def adder(self, w, v):
        return sum([i*j for i,j in zip(w,v)])

    def signum(self, val):
        if val >= 0:
            return 1
        else:
            return -1

    def scalar_prod(self, v, s):
        r = []
        for i in range(0, len(v)):
            r.append(s*v[i])
        return r
    
    def vector_adapt(self, w, x, ex, d):
        vector_sum = lambda x,y: [a+b for a,b in zip(x,y)]
        diff = self.scalar_prod(x, self.eta*(ex-d))
        updated_weights = vector_sum(w, diff)
        return updated_weights

    def train(self):
        for i in range(0, len(self.x)):
            self.x[i].insert(0,self.y[i])
        
        for i in range(0, 20):
            for j in range(0, len(self.x)):
                expected_decision = self.y[j]
                decision = self.signum(self.adder(self.weights, self.x[j]))
                if decision != expected_decision:
                    self.weights = self.vector_adapt(self.weights, self.x[j], expected_decision, decision)
                
    def classify(self, v):
        v.insert(0, 1)
        decision = self.signum(self.adder(self.weights, v))
        return decision

